Configure publishing of the email editors php package

costasovo opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR adds a GitHub workflow for mirroring the email editor's PHP package to an extra repository, which will be used for publishing the package to Packagist.

The workflow was inspired by [the workflow used for mirroring the core plugin](https://github.com/woocommerce/woocommerce/pull/52613/files).

I[ tested the workflow in a forked monorepo](https://github.com/costasovo/woocommerce/actions/workflows/mirror-php-package.yml) and here is [the mirrored package created in the test](https://github.com/costasovo/email-editor/tree/trunk).

Closes WOOPLUG-3249

Note: There will be follow-up tasks for adding usage instructions doc and workflow for preparing the release (generating changelog, etc.)

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

I don't know how to test a workflow on a branch. I [tested it in a fork](https://github.com/costasovo/woocommerce/actions/workflows/mirror-php-package.yml) and then made some minor adjustments to work with the Woo monorepo.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

costasovo committed:
Add the mirror PHP package workflow and cleanup script

WOOPLUG-3249

costasovo committed:
Rename JS config package for email editor PHP package

This is to align with the naming in the JS package.
WOOPLUG-3249

costasovo committed:
Add changelog

WOOPLUG-3249

costasovo committed:
Update readme.md in the editor PHP package

WOOPLUG-3249

costasovo commented on .github/workflows/mirror-php-package.yml:
```diff
@@ -0,0 +1,68 @@
+      - name: Push to mirror
+        uses: Automattic/action-push-to-mirrors@v2.2.2
+        with:
+          token: ${{ secrets.API_TOKEN_GITHUB }}
+          username: matticbot
```
I'm not sure about the credentials. I used the same as was in [the old mirrors workflow](https://github.com/woocommerce/woocommerce/pull/52613/files).

lysyjan reviewed (approved):
It looks good to me.

kalessil commented on .github/workflows/mirror-php-package.yml:
```diff
@@ -0,0 +1,68 @@
+name: Mirror PHP package
```
We should add more context here, particularly why we need this mirroring.

kalessil replied:
We need to ensure that prepping PHP packages for release is working. See `prepare-package-release.yml` on cli-commands: `./tools/package-release/bin/dev prepare @woocommerce/email-editor-php-config` should bump versions and handle changelog generation properly (you can test locally, it 's not committing the changes).

kalessil commented on packages/php/email-editor/tasks/mirror-prepare.sh:
```diff
@@ -0,0 +1,62 @@
+
+# List of files and directories to delete
+ITEMS_TO_DELETE=(
+    "node_modules"
```
Well, with the list so long, it makes sense to specify what to copy into the target directory.

This way, the list will require minimal maintenance in the long run and can be a bit faster as less IO is involved.

kalessil replied:
Also, rsync IIRC allows you to copy multiple sources into a single location - can simplify this script further.

kalessil commented on .github/workflows/mirror-php-package.yml:
```diff
@@ -0,0 +1,68 @@
+              uses: Automattic/action-push-to-mirrors@v2.2.2
+              with:
+                  token: ${{ secrets.API_TOKEN_GITHUB }}
+                  username: matticbot
```
We should use `woocommercebot` and grant it write permission to the mirror-repos.

kalessil commented on .github/workflows/mirror-php-package.yml:
```diff
@@ -0,0 +1,68 @@
+                    exit 1
+                  fi
+
+            - name: Setup mirror directory structure
```
From a maintenance perspective, I think it makes sense to push mirroring preps back to packages.

From this workflow, we should just run `pnpm --filter @woocommerce/email-editor-php-config build:composer-package`, which will generate `build` directory in the package itself with sources and `mirrors.txt`.

And the path for mirroring for the next step:
```bash
mirror=$( pnpm --filter @woocommerce/email-editor-php-config list --json | jq --compact-output '.[]' | jq --raw-output '( .path + "/build")' )
```

costasovo committed:
Update email editor package and tooling to support packages-prepare workflow

WOOPLUG-3249

costasovo committed:
Add build:composer-package script

WOOPLUG-3249

costasovo committed:
Update the PHP package mirror workflow to use prepared commands

WOOPLUG-3249

costasovo committed:
Add an extra readme.md for the mirrored repository

WOOPLUG-3249

costasovo committed:
Remove the unused mirror cleanup script

WOOPLUG-3249

costasovo commented:
@kalessil I believe I've addressed the comments.

In https://github.com/woocommerce/woocommerce/pull/59058/commits/2119c23d3b54392e55c4bea1b716d75a6a5ff05c I added a new formatter so that the command `./tools/package-release/bin/dev prepare @woocommerce/email-editor-config` generates a reasonable changelog. If I understand correctly, for the initial releas,e we will need to use ./tools/package-release/bin/dev prepare `./tools/package-release/bin/dev prepare @woocommerce/email-editor-config --initial-release`

In https://github.com/woocommerce/woocommerce/pull/59058/commits/7a5a362a384a1c6ad9f7506059f85faa70043080 I added a new script that copies only files we want to mirror.

In https://github.com/woocommerce/woocommerce/pull/59058/commits/63eb0192bf6aaf4806ba583eb638dd52a1f549e8 I updated the workflow.

In https://github.com/woocommerce/woocommerce/pull/59058/commits/c7a7736076f8a41de4b20194f6ee1ce385ccdb92 I added a simple MD file to replace the original README.md that points developers to Woo Monorepo, where we actually develop the package.

kalessil commented on tools/changelogger/class-php-package-formatter.php:
```diff
@@ -0,0 +1,49 @@
+	 *
+	 * @return string Link to the version's release.
+	 */
+	public function getReleaseLink( $version ) {
```
Let's add parameter and type declarations here, along with a strict type directive at the top.

kalessil commented on tools/changelogger/class-php-package-formatter.php:
```diff
@@ -0,0 +1,49 @@
+	 */
+	public function getReleaseLink( $version ) {
+		// Capture anything past /woocommerce in the current working directory.
+		preg_match( '/\/packages\/php\/(.+)/', getcwd(), $path );
```
I have mixed feelings about this one: can we instead rely to the package name from `composer.json` (fully or partially)?

kalessil commented on .github/workflows/mirror-php-package.yml:
```diff
@@ -0,0 +1,38 @@
+                  pnpm --filter "${{ github.event.inputs.package }}" build:composer-package
+
+            - name: Grab package build directory
+              run: echo "mirror_dir=$( pnpm --filter ${{ github.event.inputs.package }} list --json | jq --compact-output '.[]' | jq --raw-output '( .path + "/build")' )" >> $GITHUB_ENV
```
I think this belongs to the previous step: if we generated the sources for mirroring there, the path export belongs there.

kalessil commented on packages/php/email-editor/tasks/build-package.sh:
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+set -euo pipefail
+
+# Script to prepare email-editor package for publishing to Packagist.org
```
Nice: next level of clarity in the script, well done!

kalessil reviewed (changes requested):
Looks goooood! Leaving some final batch of comments - will do testing as we address those.

costasovo committed:
Add tyhehints to Formatter::getReleaseLink

WOOPLUG-3249

costasovo committed:
Use the package name from composer.json for creating the release link

WOOPLUG-3249

costasovo committed:
Merge the build path export to the build package step of the PHP package mirror workflow

WOOPLUG-3249

kalessil committed:
Review: fix package prepare for initial release.

kalessil committed:
Review: package build tweaks.

kalessil committed:
Review: workflow tweaks.

kalessil reviewed (approved):
LGTM! Working as expected.

Pushing some final touches in 65642e205c58445bd6b1cf93f8067b4be6a93ab3, b822a073d30ab2fe954553a8d43b288ca2320356, and ef38e2762d90c5e52b922de84bc1dbf351cd087b (minor bugs, security, cross-envs interoprability).

kalessil committed:
Review: cleanup.
